server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java

```diff
@@ -0,0 +1,282 @@
+         mkdirs(cacheRoot);
+
+         // Write configuration file
+         String xmlFileName = configFile(cacheName);
+         Path xmlPath = cacheRoot.resolve(xmlFileName);
+         try (OutputStream os = Files.newOutputStream(xmlPath)) {
+            parserRegistry.serialize(os, cacheName, configuration);
+         } catch (XMLStreamException | IOException e) {
+            throw new CacheException(String.format("Unable to create backup file '%s'", xmlFileName), e);
+         }
+
+         // Write in-memory cache contents to .dat file if the cache is not empty
+         if (cache.isEmpty())
+            return;
+
+         ComponentRegistry cr = cache.getComponentRegistry();
+         ClusterPublisherManager<?, ?> clusterPublisherManager = cr.getClusterPublisherManager().running();
+         SerializationContextRegistry ctxRegistry = cr.getGlobalComponentRegistry().getComponent(SerializationContextRegistry.class);
+         ImmutableSerializationContext serCtx = ctxRegistry.getPersistenceCtx();
+
+         String dataFileName = dataFile(cacheName);
+         Path datFile = cacheRoot.resolve(dataFileName);
+
+         Publisher<CacheEntry<?, ?>> p = s -> clusterPublisherManager.entryPublisher(null, null, null, false,
+               DeliveryGuarantee.EXACTLY_ONCE, BUFFER_SIZE, PublisherTransformers.identity())
+               .subscribe(s);
+
+         StorageConfigurationManager scm = cr.getComponent(StorageConfigurationManager.class);
+         boolean keyMarshalling = MediaType.APPLICATION_OBJECT.equals(scm.getKeyStorageMediaType());
+         boolean valueMarshalling = MediaType.APPLICATION_OBJECT.equals(scm.getValueStorageMediaType());
+         PersistenceMarshaller persistenceMarshaller = cr.getPersistenceMarshaller();
+         Marshaller userMarshaller = persistenceMarshaller.getUserMarshaller();
+         Flowable.using(
+               () -> Files.newOutputStream(datFile),
+               output ->
+                     Flowable.fromPublisher(p)
+                           .buffer(BUFFER_SIZE)
+                           .flatMap(Flowable::fromIterable)
+                           .map(e -> {
+                              CacheBackupEntry be = new CacheBackupEntry();
+                              be.key = keyMarshalling ? marshall(e.getKey(), userMarshaller) : (byte[]) scm.getKeyWrapper().unwrap(e.getKey());
+                              be.value = valueMarshalling ? marshall(e.getValue(), userMarshaller) : (byte[]) scm.getValueWrapper().unwrap(e.getKey());
+                              be.metadata = marshall(e.getMetadata(), persistenceMarshaller);
+                              be.internalMetadata = e.getInternalMetadata();
+                              be.created = e.getCreated();
+                              be.lastUsed = e.getLastUsed();
+                              return be;
+                           })
+                           .doOnNext(e -> writeMessageStream(e, serCtx, output))
```

Comment:
I don't think so, because `subscribe` returns a `Disposable` but we need to return a `Publisher` so that it can be consumed as the `Flowable.using` `sourceSupplier`. This is similar to the `JpaStore#createBatchFlowable` code.

Reply:
👍 Forgot it returns a Disposable :(